File: ServeurJavaPPIL/Monde/Vecteur2D.java
package Monde;

public class Vecteur2D {
    public double x, y;

    public Vecteur2D(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public Vecteur2D clone() {
        return new Vecteur2D(x, y);
    }

    public Vecteur2D addition(Vecteur2D u) {
        return new Vecteur2D(x + u.x, y + u.y);
    }

    public Vecteur2D multiplication(double a) {
        return new Vecteur2D(x * a, y * a);
    }

    public Vecteur2D inverse() {
        return new Vecteur2D(-x, -y);
    }

    public Vecteur2D soustraction(Vecteur2D u) {
        return new Vecteur2D(x - u.x, y - u.y);
    }

    public double norm() {
        return Math.sqrt(x * x + y * y);
    }

    public double produitScalaire(Vecteur2D u) {
        return x * u.x + y * u.y;
    }

    public double distance(Vecteur2D b) {
        return Math.sqrt(Math.pow(x - b.x, 2) + Math.pow(y - b.y, 2));
    }

    public boolean equals(Vecteur2D u) {
        return (x == u.x && y == u.y);
    }

    @Override
    public String toString() {
        return "(" + x + ", " + y + ")";
    }

    public static Vecteur2D multiplication(double a, Vecteur2D u) {
        return u.multiplication(a);
    }
}

File: ClienCpp/Formes/Vecteur2D.cpp
#include "Vecteur2D.h"

const Vecteur2D* Vecteur2D::clone() const {
    return new Vecteur2D(*this);
}

double Vecteur2D::norm() const {
    return abs(sqrt(x*x + y*y));
}

double Vecteur2D::produitScalaire(const Vecteur2D & u) const {
    return x*u.x + y*u.y;
}

Vecteur2D::operator string() const
{
    ostringstream os;
    os << "(" << x <<", " << y << ")";
    return os.str();
}

ostream & operator << (ostream & os, const Vecteur2D & u)
{
    os << (string) u;
    return os;
}